Introduce kstatus watcher

AustinAbro321 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->
closes #8661

**What this PR does / why we need it**:
This PR introduces a statusWaiter using kstatus. More details here -  [H4HIP: Wait with kstatus](https://github.com/helm/community/pull/374)

Notes to reviewers:
- I have created this repo with practical examples to show the differences between the wait logic https://github.com/AustinAbro321/helm-wait-examples. 
- kstatus is behind an interface and not directly exposed by this library, however library users can still use the new waiter by using method `SetWaiter(kube.StatusWatcherStrategy)` on a `kube.Client` object then calling `client.Wait`
- I have not written any docs, I assume that can be done when we are closer to Helm 4, a lot of it is covered by linking to - https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md

**If applicable**:
- [x] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

AustinAbro321 committed:
WIP

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
Merge branch 'refactor-wait' of github.com:AustinAbro321/helm into refactor-wait

AustinAbro321 committed:
extending factory to enable getting a watcher

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
understand it better

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
basic design up and balling

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
make a working test

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
cleanup test

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
change structure of client

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
only emebed waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
refactor new

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
refactor test

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
refactor test

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
option to wait for jobs

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
comments

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
paused as ready now working

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
paused as ready

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
context

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
simplify things

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
save state while I change up tests

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
wait for delete working

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
unknown status

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
delete wait and get completed phase

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
merge

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
go fmt

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
provide path for creating new legacy waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
status wait

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
ability to create different waiters

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
helm waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
Helm waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
status function

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
change error messages

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
general error message

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
get rid of ext interface

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
update names

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
implement logger

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
implement logger

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
introduce test for status wait

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
t.Parrallel

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
test cleanup

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
get rid of rest client

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
renames

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
refactor obj logic

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
shorten interface

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
use dynamic rest mapper

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
remove rest mapper

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
go mod tidy

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
merge

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
diff

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
simplify messages

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
go fmt

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
Merge remote-tracking branch 'origin/main' into refactor-wait

AustinAbro321 committed:
unexport newWaiter function

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

gjenkins8 commented on pkg/kube/statuswait.go:
```diff
@@ -0,0 +1,180 @@
+/*
```
If we are moving this logic, perhaps put it into its own (internal) package?

AustinAbro321 replied:
I moved the statusreaders to their own package in internal, but I think it would be a good idea to keep the waiters in kube, for this PR. We'd have to change the Waiter interface to not accept a kube.ResourceList to avoid cyclic imports. This PR is already XXL, I'd like to avoid more changes.

AustinAbro321 committed:
WIP custom status reader

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
merge main

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
switch client

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
tests passing

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
tests passing

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
tests passing

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
tests passing

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
tests passing

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
start watch until ready

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
custom status readers look good

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
tests for status reader

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
pods and jobs working

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
move statusreaders to it's own package

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
lint

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
Merge branch 'main' into refactor-wait

AustinAbro321 committed:
comment

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
set command line flags

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
add hook only waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
lint

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
set waiter in functions

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
wait for delete

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
lint

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
better name

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
print

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
better help text

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
refactor

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
refactor

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
test refactoring

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
merge

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
cleanup

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
fmt

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
add back interface log check

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
merge

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
merge

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
Merge branch 'main' into refactor-wait

AustinAbro321 committed:
Merge branch 'main' into refactor-wait

scottrigby commented on internal/statusreaders/job_status_reader.go:
```diff
@@ -0,0 +1,120 @@
+
+package statusreaders
+
+// This file was copied and modified from https://github.com/fluxcd/kustomize-controller/blob/main/internal/statusreaders/job.go
+
```
```suggestion
```

scottrigby commented on internal/statusreaders/job_status_reader.go:
```diff
@@ -0,0 +1,120 @@
+/*
+Copyright The Helm Authors.
```
Apache license requires us to retain the Flux copyright on these files.

The fluxcd/kustomize-controller project does not have a NOTICE file, so it's sufficient to add the Helm copyright to the existing Flux license header on this and any other derived files.

```suggestion
Copyright The Helm Authors.
This file was initially copied and modified from
    https://github.com/fluxcd/kustomize-controller/blob/main/internal/statusreaders/job.go
Copyright 2022 The Flux authors
```

scottrigby commented on internal/statusreaders/job_status_reader_test.go:
```diff
@@ -0,0 +1,114 @@
+/*
+Copyright The Helm Authors.
```
Same as above. Also fixing link to original test file here:

```suggestion
Copyright The Helm Authors.
This file was initially copied and modified from
    https://github.com/fluxcd/kustomize-controller/blob/main/internal/statusreaders/job_test.go
Copyright 2022 The Flux authors
```

scottrigby reviewed (changes requested):
still reviewing, but first off we need to retain the existing license headers from the two files drawn from Flux's statusreaders package.

AustinAbro321 committed:
Apply suggestions from code review

Co-authored-by: Scott Rigby <scott@r6by.com>
Signed-off-by: Austin Abro <37223396+AustinAbro321@users.noreply.github.com>

AustinAbro321 committed:
hook only strategy when wait=false

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

scottrigby reviewed (changes requested):
Summary: revert back to flux packages, and vendor `sigs.k8s.io/controller-runtime/pkg/client/apiutil`.

scottrigby commented:
needs rebase since https://github.com/helm/helm/pull/30618 was merged

AustinAbro321 committed:
Merge branch 'main' into refactor-wait

AustinAbro321 committed:
Update internal/statusreaders/job_status_reader.go

Co-authored-by: Scott Rigby <scott@r6by.com>
Signed-off-by: Austin Abro <37223396+AustinAbro321@users.noreply.github.com>

AustinAbro321 committed:
restmapper

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
rest mapper

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 commented on pkg/kube/client.go:
```diff
@@ -27,34 +27,28 @@ import (
 	apiextv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
```
Added the restmapper in `internal/client`. LMK if there is a better place to put that, and thank you for the detailed context and history.

alvaroaleman replied:
To port over some context from https://github.com/kubernetes-sigs/controller-runtime/issues/3133, this issue exists a lot longer than 0.19 and is completely unrelated to aggregated discovery.

AustinAbro321 committed:
copyright things

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

scottrigby reviewed (dismissed):
All feedback was implemented, and this e2e test repo—which I checked closely https://github.com/AustinAbro321/helm-wait-examples—shows this PR working well.  LGTM 👍 Other reviewers, please see my review comments for reasoning and context on the fixes.

matheuscscp reviewed (approved):
LGTM

Should still wait for @stefanprodan's review before merging

AustinAbro321 committed:
merge main

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

gjenkins8 commented on pkg/kube/wait.go:
```diff
@@ -31,25 +32,43 @@ import (
-	log     func(string, ...interface{})
+// HelmWaiter is the legacy implementation of the Waiter interface. This logic was used by default in Helm 3
+// Helm 4 now uses the StatusWaiter implementation instead
+type HelmWaiter struct {
```
Can we make this type privte (regardless of name). Also suggest 'legacyWaiter` over `helmWaiter`, as that is more specific / `helmWaiter` could easily describe e.g. the default waiter 

```suggestion
type legacyWaiter struct {
```

gjenkins8 reviewed (commented):
made an initial pass. biggest concern, is it looks like we are modifying the global cfg object. other things mostly nits

need to make a second pass to that looks at some of the logic in depth

AustinAbro321 committed:
client->restmapper

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 commented on internal/client/client.go:
```diff
@@ -0,0 +1,372 @@
+/*
```
re-named to rest-mapper

AustinAbro321 committed:
cleanup if statement

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 commented on pkg/action/upgrade.go:
```diff
@@ -155,7 +155,14 @@ func (u *Upgrade) RunWithContext(ctx context.Context, name string, chart *chart.
 	// Make sure if Atomic is set, that wait is set as well. This makes it so
 	// the user doesn't have to specify both
-	u.Wait = u.Wait || u.Atomic
+	if u.Wait == kube.HookOnlyStrategy {
```
I don't feel strongly, I'll change them to doing it this way

AustinAbro321 committed:
Update pkg/cmd/flags.go

Co-authored-by: George Jenkins <gvjenkins@gmail.com>
Signed-off-by: Austin Abro <37223396+AustinAbro321@users.noreply.github.com>

AustinAbro321 committed:
warnings

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
legacy waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

stefanprodan commented:
I've tested the RESTMapper from https://github.com/kubernetes-sigs/controller-runtime/pull/3151 and the issue with preferred version has been fixed. I suggest waiting on @alvaroaleman PR to be merged and backported to controller-runtime v0.20. Then we can drop the RESTMapper vendoring from this PR and use upstream.

AustinAbro321 committed:
go mod tidy

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 commented on pkg/action/rollback.go:
```diff
@@ -60,6 +61,10 @@ func (r *Rollback) Run(name string) error {
 		return err
 	}
 
+	if err := r.cfg.KubeClient.SetWaiter(r.Wait); err != nil {
```
Makes sense, I'll try that out

AustinAbro321 replied:
Went with a strategy like above, lmk what you think

scottrigby replied:
`GetWaiter` looks to be addressed in this commit https://github.com/helm/helm/pull/13604/commits/386523bdbc6f5e5f289ade7d9d4cf4c935354450.

AustinAbro321 committed:
update to get waiter instead of set

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
merge

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
switch back to k8s rest mapper

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
go fmt

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 commented:
@stefanprodan I added dependency `sigs.k8s.io/controller-runtime v0.20.4` which includes the linked PR. Began using `apiutil.NewDynamicRestMapper` from controller-runtime and deleted the RESTMapper vendoring

scottrigby reviewed (approved):
LGTM 👍 

Same reasoning as here https://github.com/helm/helm/pull/13604#pullrequestreview-2665030455, but also reviewed how feedback from @gjenkins8 was addressed 👍 and also excellent that was landed in upstream sigs.k8s.io/controller-runtime as pointed out by @stefanprodan and implemented in this PR. 

This is a large PR for Helm, but something Helm maintainers have wanted to see in Helm for a while. This is extremely well done, and given our Helm 4 schedule I think this is our best opportunity to get kstatus support into Helm.

Also special thanks to @stefanprodan for the thorough review—this feedback is invaluable since Flux figured out how to properly add kstatus support as a Helm SDK user. It's worth noting that this kstuatus support approach has been long since thoroughly tested by Helm users via Flux.

sabre1041 reviewed (approved):
Verified test cases
